docs/controllers/boshdeployment.md

```diff
@@ -0,0 +1,79 @@
+# BOSHDeployment
+
+- [BOSHDeployment](#bosheployment)
+  - [Description](#description)
+  - [Reconcilers](#reconcilers)
+    - [Deployment Reconciler](#deployment-reconciler)
+    - [Generated Variable Reconciler](#generated-variable-reconciler)
+    - [BPM Reconciler](#bpm-reconciler)
+
+## Description
+
+A BOSH deployment is created from a deployment manifest and optionally ops files.
+
+The deployment manifest is based on a vanilla BOSH deployment manifest,
+The ops files modify the deployment manifest. For example, ops files can be used to replace release tarballs with [docker images](https://ci.flintstone.cf.cloud.ibm.com/teams/containerization/pipelines/release-images), thus enabling deployment on Kubernetes.
+
+A deployment is represented by the `boshdeployments.fissile.cloudfoundry.org` (`bdpl`) custom resource, defined in [`boshdeployment_crd.yaml`](https://github.com/cloudfoundry-incubator/cf-operator/tree/master/deploy/helm/cf-operator/templates/fissile_v1alpha1_boshdeployment_crd.yaml).
+This [bdpl custom resource](https://github.com/cloudfoundry-incubator/cf-operator/tree/master/docs/examples/bosh-deployment/boshdeployment.yaml) contains references to config maps or secrets containing the actual manifests content.
+
+After creating the bdpl resource on kubernetes, i.e. via `kubectl apply`, the CF operator will start the reconciliation, which will [eventually result in the deployment](https://docs.google.com/drawings/d/126ExNqPxDg1LcB14pbtS5S-iJzLYPyXZ5Jr9vTfFqXA/edit?usp=sharing) of the BOSH release on Kubernetes.
+
+## Reconcilers
+
+### Deployment Reconciler
+
+Watches for:
+
+- `BOSHDeployment`
+- `ConfigMap`/`Secret` for ops files and the deployment manifest
+- `Pods` created for the BOSHDeployment
+
+Creates/updates:
+
+- the "Variable Interpolation" [auto-errand extended job](https://github.com/cloudfoundry-incubator/cf-operator/tree/master/docs/controllers/extendedjob.md##one-off-jobs-auto-errands)
+- the "Data Gathering" auto-errand extended job
+- the "with-ops" secret
+
+> **Note:** the output of the ["Variable Interpolation"](https://github.com/cloudfoundry-incubator/cf-operator/tree/master/docs/commands/cf-operator_util_variable-interpolation.md) `ExtendedJob` is the input for the "Data Gathering" `ExtendedJob`.
+> The ["Data Gathering"](https://github.com/cloudfoundry-incubator/cf-operator/tree/master/docs/commands/cf-operator_util_data-gather.md) step generates 2 secrets for each Instance Group:
+>
+> - Instance Group Resolved Properties
+> - Instance Group BPM
+>
+> The "Versioned" `Secret` for Instance Group Resolved Properties is referenced by the Instance Group `ExtendedStatefulSets` and `ExtendedJobs`
+
+Ownership:
+
+TODO document if this will have to keep ownership on something?
```

Comment:
Ownership can be used in cleanup phase, so it is needed. AFAIK

Reply:
Based on docs, ownership is used for garbage collection. As a general rule, all resources we _create_ to reconcile a specific object should be owned by that object.

Do we use it for anything else, and if garbage collection is automatic, why do we still need finalizers?

Reply:
Finalizers are mostly used to remove ownership from objects we don't want to clean up: custom resources created by the user for example.

So when you register ownership on a 'foreign' pod, you also set a finalizer on the new owner. When deleting the owner you can remove finalizer from the owner and the ownership from the pod.

Reply:
I thought we use ownership to make it happens that when you delete an instance, the instance's dependents(sts, jobs, services, etc) are also deleted. And `finalizers` only controls this process.

1. Without `finalizers`, this deletion is automated
2. Without ownership, this deletion doesn't happen and kube leaves orphans to us